Improve client b1 to implement 1

schettn opened the pull request:
**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?


**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [ ] Documentation (non-breaking change which adds documentation)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)


**What is the current behavior? (link to any open issues here)**
- Currently there is one method which handles all Apollo types. #37 
- Currently there are plenty of interfaces. #38 
- Currently it is not possible to perform a custom session aware task. #39 
- Currently the clients Apollo endpoint notation is misleading. #40 


**What is the new behavior (if this is a feature change)?**
- Now there are separate methods for each Apollo type. Closes #37 
- Now the number of interfaces has been removed by adding type parameter handling to Apollo endpoint. Closes #38 
- Now there it is possible to perform custom session aware tasks. Closes #39 
- Now the notation of the Apollo endpoint of clients has been adjusted. Currently the old version is still present but will be removed in a upcoming release. #40


**Other information**:
- Ref: 🐍

schettn committed:
Split Apollo send functionallity

Now there are two separate methods for mutations and queries.
The Apollo endpoint can handle type parameters now.

schettn committed:
Update SNEK template

The tasks have been simplified and all "Response" interfaces
have been removed because they are no longer required due
to the previous Apollo endpoint adjustments.

schettn committed:
Remove unnecessary interfaces

Some unnecessary interfaces have been removed.

schettn committed:
Add token getter

Now it is possible to the the token from the super class via
a getter.

schettn committed:
Adapt sessions to new Apollo  

Now Apollo is called via the newly implemented methods.

schettn committed:
Adapt sessions to task responses

Due to the change in the structure of the task responses the sessions have been adapted.

schettn committed:
Improve some documentation

Documentation has been improved to fit to the code functionality. But of course also to improve the style. :)

schettn committed:
Add custom task

With the custom task it is possible to run a query or mutation
within a session. Therefore latest token and refresh token are 
set by default.

schettn committed:
Change endpoint field to gql

Now it is clear what a field does by its name. A specific endpoint should be never called "endpoint" but the type name 
of a endpoint.
To be non-breaking the old field name is present but will be removed in the upcoming release.

pinterid commented on src/endpoints/apollo.ts:
```diff
@@ -75,52 +78,75 @@ class Apollo implements ApolloEndpoint {
+      console.warn(
+        "No query type specified! Please re-check." +
+          "Selecting type 'query' as default"
+      );
```
Missing newline.

pinterid commented on src/endpoints/index.ts:
```diff
@@ -5,6 +5,12 @@
+import { FetchResult } from "apollo-link";
+import { ApolloQueryResult } from "apollo-client";
+
+export type ApolloResult<T> =
```
Put export to the bottom of the file.

pinterid commented on src/session/sessions.ts:
```diff
@@ -168,16 +177,16 @@ class SnekSession extends Session {
    *
-   * @param {string} user A User defined by username and password
-   * @returns {Promise<UserData>} A UserData object
+   * @param {string} user A User defined by username and password.
```
Remove dot.

pinterid commented on src/session/sessions.ts:
```diff
@@ -168,16 +177,16 @@ class SnekSession extends Session {
-   * @param {string} user A User defined by username and password
-   * @returns {Promise<UserData>} A UserData object
+   * @param {string} user A User defined by username and password.
+   * @returns {Promise<any>} A UserData object.
```
Remove dot.

pinterid commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -1,101 +1,70 @@
 //#endregion
 
 //#region > Classes
 /** @class A set of session aware Tasks */
```
Write "Tasks" in lower case.

pinterid commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -1,101 +1,70 @@
    * Anonymous login.
    *
-   * @returns {Promise<AuthResponse>} A JWT token
+   * @returns {Promise<ApolloResult<AuthData>>} A JWT token.
```
Remove dot and is "token" needed in this case?

pinterid commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -109,64 +78,55 @@ class SnekGqlAuthTasks extends TaskError {
    * @param {string} user A User defined by username and password
-   * @returns {Promise<AuthData>} A JWT token
    * @description Authenticates a real user to obtain a JWT
+   * @returns {Promise<ApolloResult<AuthData>>} A JWT token
```
Is "token" required here?

pinterid commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -109,64 +78,55 @@ class SnekGqlAuthTasks extends TaskError {
    *
    * @param {string} user A User defined by username and password
-   * @returns {Promise<RefreshResponse>} A JWT token
+   * @returns {Promise<ApolloResult<RefreshData>>} A JWT token
```
Is "token" required here?

pinterid commented on src/templates/snek/gql/tasks/general.ts:
```diff
@@ -1,99 +1,76 @@
- * @interface GitlabServerData defines the structure of the specific data a
- *                             GitlabServerResponse contains.
+ * @interface GitlabServerData defines the structure of the Gitlab server result
+ *                             data
```
Missing dot.

pinterid commented on src/templates/snek/gql/tasks/general.ts:
```diff
@@ -1,99 +1,76 @@
    * @author Nico Schett <contact@schett.net>
-   * @param {string} session A session for the tasks
+   * @param {string} parent The parent task
+   * @param {string} template A template set containing mutation and queries
```
Use the plural of "mutation"

schettn replied:
Tis param should be removed totally.

pinterid commented on src/templates/snek/gql/tasks/index.ts:
```diff
@@ -38,15 +38,40 @@ class SnekTasks {
+
+    this.template = session.template.snekGql;
+
+    /* init tasks */
```
Change "init" to "Init"

pinterid commented on src/templates/snek/gql/tasks/index.ts:
```diff
@@ -38,15 +38,40 @@ class SnekTasks {
+      console.warn(
+        "No query type specified! Please re-check." +
+          "Selecting type 'query' as default"
+      );
```
Missing newline.

pinterid commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -73,116 +43,98 @@ interface ProfileData {
- * @interface WhoamiData defines the structure of the specific data a
- *                       whoami response contains.
+ * @interface WhoamiData defines the structure of the whoami result
+ *                        data.
```
Remove space before "data".

pinterid commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -73,116 +43,98 @@ interface ProfileData {
    * @author Nico Schett <contact@schett.net>
-   * @param {string} session A session for the tasks
+   * @param {string} parent The parent task
+   * @param {string} template A template set containing mutation and queries
```
Use the plural of "mutation"

schettn replied:
This should be removed.

pinterid commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -73,116 +43,98 @@ interface ProfileData {
    *
    * @param {object} values Registration parameters like username, email,...
-   * @returns {Promise<RegistrationResponse>} A JWT token
+   * @returns {Promise<ApolloResult<RegistrationData>>} A JWT token
```
Is "token" required?

pinterid commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -73,116 +43,98 @@ interface ProfileData {
    *
    * @param {string} platformData A serialized JSON object to be cached
-   * @returns {Promise<CacheResponse>} A JWT token
+   * @returns {Promise<ApolloResult<CacheData>>} A JWT token
```
Is "token" required?

pinterid reviewed (changes requested):
Looks good after some changes.

schettn committed:
Adjust interface names

All interface which does not have a class with the same name
does not need a capital I.

schettn committed:
Improve documentation quality

The quality of the documentation has been improved due to
a request of @pinterid.

Aichnerc reviewed (approved):
LGTM 🎁

schettn committed:
Remove unnecessary console error

A unwanted console error is removed now.

pinterid reviewed (approved):
Looks good to merge.
